From 909b329597cb8ba3dc7ce750ae8f3083813a8d16 Mon Sep 17 00:00:00 2001
From: Kazunori Kobayashi <kkobayas@igel.co.jp>
Date: Wed, 22 Aug 2012 12:19:42 +0900
Subject: [PATCH 50/97] gst/h264parse: only check out of order timestamps on
 baseline profile streams

h264parse checks and corrects out of order timestamps, but
non-baseline profile streams can have B slices that could be
reordered. This patch disables timestamp checks for non-baseline
profile streams.
---
 gst/h264parse/gsth264parse.c |    6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/gst/h264parse/gsth264parse.c b/gst/h264parse/gsth264parse.c
index bc58269..7dfd2b7 100644
--- a/gst/h264parse/gsth264parse.c
+++ b/gst/h264parse/gsth264parse.c
@@ -2168,10 +2168,14 @@ gst_h264_parse_chain_forward (GstH264Parse * h264parse, gboolean discont,
 
       /* Ignore upstream dts that stalls or goes backward. Upstream elements
        * like filesrc would keep on writing timestamp=0.  XXX: is this correct?
+       * Only the dts in the baseline profile should be ignored when a timestamp
+       * goes backward because other profiles can have B slices that could be
+       * reordered in their own stream.
        * TODO: better way to detect whether upstream timstamps are useful */
       if (h264parse->last_outbuf_dts != GST_CLOCK_TIME_NONE
           && outbuf_dts != GST_CLOCK_TIME_NONE
-          && outbuf_dts <= h264parse->last_outbuf_dts)
+          && outbuf_dts <= h264parse->last_outbuf_dts
+          && h264parse->sps && h264parse->sps->profile_idc == 66)
         outbuf_dts = GST_CLOCK_TIME_NONE;
 
       if ((got_frame || delta_unit) && start) {
-- 
1.7.10.4

